### PR TITLE
LibCrypto: Add a minimal DER encoder

### PR DESCRIPTION
**Pull request by alimpfard**
Progress towards #23562.

Todo:
- Add some stuff to encode RSAPrivateKey, etc.

### CODE REVIEW

**Comment by alimpfard** [comment]
(WIP still)

File: Userland/Libraries/LibCrypto/PK/RSA.h [hunk 1]
```diff
@@ -65,21 +96,41 @@ class RSAPrivateKey {
...
 
 private:
     Integer m_modulus;
     Integer m_private_exponent;
     Integer m_public_exponent;
+    Integer m_p;
+    Integer m_q;
+    Integer m_dp;   // d mod (p-1)
+    Integer m_dq;   // d mod (q-1)
+    Integer m_qinv; // q^-1 mod p
```
Comment (ADKaster) [inline Userland/Libraries/LibCrypto/PK/RSA.h:133]: should we save these as the names they have in the interface/the RSAPrivateKey ASN.1 thingy, and keep the RSA-spec-names as comments?

File: Userland/Libraries/LibCrypto/PK/RSA.h [hunk 2]
```diff
@@ -65,21 +96,41 @@ class RSAPrivateKey {
...
-    void set_length(size_t length) { m_length = length; }
 
-    void set(Integer n, Integer d, Integer e)
+    ErrorOr<ByteBuffer> export_as_der() const
     {
-        m_modulus = move(n);
-        m_private_exponent = move(d);
-        m_public_exponent = move(e);
-        m_length = m_modulus.trimmed_length() * sizeof(u32);
+        ASN1::Encoder encoder;
```
Comment (ADKaster) [inline Userland/Libraries/LibCrypto/PK/RSA.h:108]: should this API take an encoder by ref, similar to how our IPC::Encoder works? That way a caller can avoid making a bunch of temp byte buffers, and simply get the byte buffer for the top level item that they want.

File: Userland/Libraries/LibCrypto/BigInt/UnsignedBigInteger.cpp [hunk 3]
```diff
@@ -550,6 +550,20 @@ void UnsignedBigInteger::set_bit_inplace(size_t bit_index)
...
 }
 
+size_t UnsignedBigInteger::count_trailing_zeroes() const
+{
+    size_t count = 0;
+    for (size_t i = 0; i < length(); ++i) {
+        if (m_words[i] == 0) {
+            count += BITS_IN_WORD;
+        } else {
+            count += __builtin_ctz(m_words[i]);
```
Comment (ADKaster) [inline Userland/Libraries/LibCrypto/BigInt/UnsignedBigInteger.cpp:560]: we have a wrapper for this in AK/BuiltinWrappers.h

File: Userland/Libraries/LibCrypto/PK/RSA.h (hunk 1 above)
Comment (alimpfard) [reply Userland/Libraries/LibCrypto/PK/RSA.h:133]: I don't really mind either, I'll rename to the spec names.

File: Userland/Libraries/LibCrypto/PK/RSA.h (hunk 2 above)
Comment (alimpfard) [reply Userland/Libraries/LibCrypto/PK/RSA.h:108]: Nothing embeds this object directly, you'll have to make a temp bytebuffer anyway...?

Comment (ADKaster) [reply Userland/Libraries/LibCrypto/PK/RSA.h:108]: SubjectPublicKeyInfo and PrivateKeyInfo do, no?

File: Userland/Libraries/LibCrypto/ASN1/DER.cpp [hunk 4]
```diff
@@ -234,6 +234,212 @@ ErrorOr<void> Decoder::leave()
...
+{
+    if (value < 0x80)
+        return write_byte(value);
+
+    size_t size = 0;
+    size_t temp = value;
+    while (temp) {
+        temp >>= 8;
+        size++;
+    }
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibCrypto/ASN1/DER.cpp:269]: This sounds like `ceil_div(ceil_log2(value, 3))`, not sure about both `ceils`

File: AK/StdLibExtraDetails.h [hunk 5]
```diff
@@ -602,6 +602,9 @@ struct EquivalentFunctionTypeImpl<T (C::*)(Args...) const> {
 template<typename Callable>
 using EquivalentFunctionType = typename EquivalentFunctionTypeImpl<Callable>::Type;
 
+template<typename T, typename... Args>
+inline constexpr auto IsAnyOf = (IsSame<T, Args> || ...);
+
```
Comment (Hendiadyoin1) [inline AK/StdLibExtraDetails.h:607]: We already have `IsOneOf`, although `Any` sounds nicer and is closer to other library functions
So maybe a separate commit renaming it

File: Userland/Libraries/LibCrypto/PK/RSA.h (hunk 2 above)
Comment (alimpfard) [reply Userland/Libraries/LibCrypto/PK/RSA.h:108]: Not directly, they embed an octet/bit string with the DER-encoded contents of this thing

File: AK/StdLibExtraDetails.h (hunk 5 above)
Comment (alimpfard) [reply AK/StdLibExtraDetails.h:607]: oh that's why I couldn't find it lol

File: Userland/Libraries/LibCrypto/ASN1/DER.cpp [hunk 6]
```diff
@@ -234,6 +235,206 @@ ErrorOr<void> Decoder::leave()
     return {};
 }
 
+ErrorOr<void> Encoder::write_tag(Class class_, Type type, Kind kind)
+{
+    auto class_byte = static_cast<u8>(class_);
+    auto type_byte = static_cast<u8>(type);
+    auto kind_byte = static_cast<u8>(kind);
```
Comment (ADKaster) [inline Userland/Libraries/LibCrypto/ASN1/DER.cpp:242]: if these are enums, prefer to_underlying, while specifying the size as u8 at the decl?

**Review by ADKaster — APPROVED** [review]
One nitpick, o.w. lgtm

**Comment by ADKaster** [comment]
... also one of your primality tests failed on serenity w/gcc https://github.com/SerenityOS/serenity/actions/runs/8280828043/job/22658122570?pr=23570#step:30:481

**Pull request merged**